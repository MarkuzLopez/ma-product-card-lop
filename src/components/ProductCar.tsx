import React, { createContext, ReactElement, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import { Product, ProductContextProps, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';


export const ProductContext =  createContext({} as  ProductContextProps);

const { Provider } = ProductContext;

export interface Props { 
    product: Product;
    //children?: ReactElement | ReactElement[];
    children: (argument: ProductCardHandlers) => ReactElement;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: onChangeArgs ) => void;
    value?: number;
    initialValue?: InitialValues;
}

export const ProductCar = ({ product, children, className, style, onChange, value, initialValue }: Props) => {

    const { counter, increaseBy, reset, isMaxCountReached  } = useProduct({ onChange, product, value, initialValue })

    const  maxCount = initialValue?.maxCount || 0
    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount    
        }}> 
        <div className={`${styles.productCard} ${className}`} style={style}>
            { children({
                count: counter,
                isMaxCountReached,
                maxCount,
                product,
                increaseBy,
                reset
            }) 
            }
        </div>
        </Provider>
    )
}
