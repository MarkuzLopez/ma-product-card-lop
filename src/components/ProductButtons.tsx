import React, { useContext } from "react";
import { ProductContext } from "./ProductCar";
import styles from '../styles/styles.module.css';
import { useCallback } from "react";

export interface Props {
    className?: string;
}

export const ProductButtons = ({className }: Props) => {
    // TODO maxCount
    const { increaseBy, counter, maxCount } = useContext(ProductContext);

    const isMaxReached = useCallback(
      // si no existe maxCount siempre va a regresar un booleano
      () => !!maxCount && counter === maxCount,
      [counter, maxCount]
    )

    // TODO:  isMaxReached callback 
    return (
        <div className={`${styles.buttonsContainer} ${className}`}>
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)} > - </button>
            <button className={styles.countLabel}>{counter} </button>
            <button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} disabled={isMaxReached()} onClick={() => increaseBy(+1)}> + </button>
        </div>
    )
}