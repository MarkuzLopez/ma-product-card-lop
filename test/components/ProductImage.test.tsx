import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCar } from '../../src/components/ProductCar';
import { product2 } from '../data/product';
import { ProductImage } from '../../src/components/ProductImage';

describe('ProductImage', () => { 

    test('debe de mostrar el componente correctamente con el titulo personalizado', () => {

        const wrapper = renderer.create(
            <ProductImage img="hhtps://hola.jpg" />
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
        //console.log(wrapper.toJSON() )
     });

     test('debe de mostrar el componente, con la imagen del producto', () => { 
        
        const wrapper = renderer.create(
            <ProductCar product={product2} >
                {
                    () => (
                        <ProductImage  />
                    )
                }
            </ProductCar>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
      });
})