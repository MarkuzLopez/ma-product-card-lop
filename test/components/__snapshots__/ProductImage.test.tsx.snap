// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostrar el componente correctamente con el titulo personalizado 1`] = `
<img
  alt="ImagePpp"
  className="productImg undefined"
  src="hhtps://hola.jpg"
/>
`;

exports[`ProductImage debe de mostrar el componente, con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="ImagePpp"
    className="productImg undefined"
    src="./coffe-mug.png"
  />
</div>
`;
